### PR TITLE
Attribution methods for API-accessible LLMs

### PR DESCRIPTION
**Pull request by sebastian-sosa**
- `attribution_metrics.py` implements attribution metrics that can be used without having model gradients.
- `token_perturbation.py` implements token replacement methods that allow to modify a prompt.
- `ExperimentLogger` tracks experiment metadata: initial prompts, model outputs, prompt perturbations, model outputs for changed prompts, attribution scores for sentences and individual tokens, etc.
- `api_llm_attribution.ipynb` shows sample usage

To run `api_llm_attribution.ipynb`, add an `OPENAI_API_KEY` variable to your `./.env` file.

### CODE REVIEW

**Review by robbiemccorkell — APPROVED** [review]
I know this is in Draft, but looking very tidy! At least good for merge into the research folder.

I'll let you guys figure out of there is more today on any particular approach once Jessica has reviewed

File: attribution/token_perturbation.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
+from typing import List
+
+import numpy as np
+from sklearn.neighbors import NearestNeighbors
+from transformers import PreTrainedTokenizer
+
+FIXED_REPLACEMENT_TOKEN = "the"
```
Comment (jessicarumbelow) [inline attribution/token_perturbation.py:7]: I think a space " " or some other more neutral token would be a better choice?

File: research/api_llm_attribution.ipynb
Comment (jessicarumbelow) [inline research/api_llm_attribution.ipynb:1]: I think client.chat.completions.create should use temperature=0.0 and seed=0

image
Weird bug here - the output is "13 minutes" but the attribution matrix logs 'the clock' as output tokens

File: attribution/attribution_metrics.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
+import math
+from typing import List, Tuple
+
+import numpy as np
+import openai
+from sklearn.metrics.pairwise import cosine_similarity
+from transformers import PreTrainedModel, PreTrainedTokenizer
+
+
+def token_prob_difference(
```
Comment (jessicarumbelow) [inline attribution/attribution_metrics.py:10]: I think this will work much better with nearest perturbation, since currently it collapses to near 0 and near 1.

We could also neatly modify this to look at the probability diff regardless of location in the output, right? For each token in the initial unperturbed output, if that token is in the perturbed output or *any* of their top 20 log probs for *any* perturbed output token, use that value. Does this make sense? That way it doesn't matter at all if the output length or position of the output token changes. (Though I can imagine this getting a bit tricky if tokens are repeated in the output).

File: attribution/attribution_metrics.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        [top_logprob.token for top_logprob in token_content.top_logprobs]
+        for token_content in perturbed_logprobs.content
+    ]
+
+    # Token displacement for each initially predicted token
+    displacement_per_token = np.zeros(len(initial_tokens))
+    MAX_TOKEN_DISPLACEMENT = 20
+    for i, token in enumerate(initial_tokens):
+        if i < len(perturbed_top_tokens) and token in perturbed_top_tokens[i]:
+            displacement_per_token[i] = abs(i - perturbed_top_tokens[i].index(token))
```
Comment (jessicarumbelow) [inline attribution/attribution_metrics.py:66]: Not sure about `i - perturbed_top_tokens[i]` - is that right?

File: attribution/token_perturbation.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+def get_increasingly_distant_token_ids(
+    token_id: int,
+    embeddings: np.ndarray,
+    n_tokens: int = 1,
+) -> List[int]:
+    if n_tokens > 4:
+        raise ValueError("n_tokens cannot be more than 4")
+
+    # Define the sequence of positions
+    positions = [10**i for i in range(1, n_tokens + 1)]
```
Comment (jessicarumbelow) [inline attribution/token_perturbation.py:79]: what’s the rationale behind get tokens at positions [10, 100, 1000, 10000]? Seems to work, just curious

File: research/api_llm_attribution.ipynb
Comment (jessicarumbelow) [reply research/api_llm_attribution.ipynb:1]: Time logging per method too please

File: attribution/token_perturbation.py
Comment (jessicarumbelow) [inline attribution/token_perturbation.py:1]: To speed things up, we could also add a perturbation method that perturbs whole *words* or *n-grams* rather than tokens, which should be both faster and easier for humans to understand. This can work the exact same way as the existing implementations, but just perturb 2 or 3 tokens etc at the same time. (For fixed perturbation, probably it makes sense to replace the whole word or n-gram with a single fixed token, rather than a single fixed token x 2 or 3 or whatever)

**Review by jessicarumbelow — CHANGES_REQUESTED** [review]
Comments inline :)

File: attribution/experiment_logger.py [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+            sentence_data.update(
+                {f"token_{i+1}": token_attr for i, token_attr in enumerate(token_attrs)}
+            )
+
+            sentences.append(sentence_data)
+
+        df_sentences = pd.DataFrame(sentences)
+        self.pretty_print(df_sentences)
+
+    def print_attribution_matrix(self, exp_id: int, attribution_strategy: str):
```
Comment (jessicarumbelow) [inline attribution/experiment_logger.py:149]: I think this would be easier if there was an option to take the row index, instead of having to specify the experiment number and attribution strategy. Or, if just the experiment number is provided, print all.

File: attribution/attribution_metrics.py (hunk 2 above)
Comment (sebastian-sosa) [reply attribution/attribution_metrics.py:10]: Yeah I thought about that, but repeated tokens may make this tricky. Common tokens will probably be repeated in the top 20 logprobs across the sentence but for unrelated reasons, artificially pushing the similarity score. 

We could also define a "window length" where to look for tokens, but I can imagine ways you can rephrase a sentence to have a very similar meaning but by shifting a token which were at the beginning of the sentence to the end. Looking at "nearby" tokens instead of the "same position" tokens wouldn't be robust in these cases.

File: attribution/attribution_metrics.py (hunk 3 above)
Comment (sebastian-sosa) [reply attribution/attribution_metrics.py:66]: You're right, thanks! 

The token displacement should just be `perturbed_top_tokens[i].index(token)`

File: attribution/token_perturbation.py (hunk 4 above)
Comment (sebastian-sosa) [reply attribution/token_perturbation.py:79]: The idea is to sort all the vocabulary by similarity, and then get exponentially more distant tokens. 

Initially, we thought about doing multiple perturbations per token. In this case, it makes sense that the multiple perturbations are exponentially more distant, as you don't want to change a token with the 2nd, 3rd, 4th, and 5th more distant tokens as those would be quite similar to the original one.

We are now just using this method to get a `distant` token - the 10.000th most distant token:
```
def get_replacement_token(
    token_id_to_replace: int,
    perturbation_strategy: str,
    embeddings: np.ndarray = None,
    tokenizer: PreTrainedTokenizer = None,
) -> int:
    if perturbation_strategy == "fixed":
        return tokenizer.encode(FIXED_REPLACEMENT_TOKEN, add_special_tokens=False)[0]
    elif perturbation_strategy == "distant":
        return get_increasingly_distant_token_ids(
            token_id_to_replace, embeddings, n_tokens=4
        )[-1]
...
```        
because of `n_tokens=4` and `[-1]`

Comment (sebastian-sosa) [reply attribution/token_perturbation.py:79]: The issue with doing _multiple_ perturbations per token is that you introduce a second dimension across which "attribution" can be computed: do you compute an attribution score by looking at how different two outputs are, by looking at "how different" the replaced token is, or an average of the two?

File: attribution/token_perturbation.py
Comment (sebastian-sosa) [reply attribution/token_perturbation.py:1]: Sounds good. Should we use stride `n`? As in, replace n tokens for a single token, measure perturbation, move `n` positions ahead and do another perturbation.

This would reduce the number of perturbations but I want to call out that the method would be susceptible to the way we group the tokens into n-grams.

File: attribution/experiment_logger.py (hunk 5 above)
Comment (sebastian-sosa) [reply attribution/experiment_logger.py:149]: One option is to ask the `exp_id` and make the `attribution_strategy` optional, and default to print another one. I'll make that.

Another option would be to ask the `row_id` instead, but that would need to be a different method. Let me know if you're interested in having that as well!

File: research/api_llm_attribution.ipynb
Comment (sebastian-sosa) [reply research/api_llm_attribution.ipynb:1]: Will set temperature and seed, thanks.
Added time logging and shared results.
Will try to reproduce the matrix bug.

File: attribution/attribution_metrics.py (hunk 2 above)
Comment (jessicarumbelow) [reply attribution/attribution_metrics.py:10]: Mmm. It's tricky. Let's leave it for now - I'll put a ticket in the backlog and maybe we can figure out an elegant way to handle it later.

File: attribution/token_perturbation.py
Comment (jessicarumbelow) [reply attribution/token_perturbation.py:1]: Yeah, I completely agree. For now, let's just implement per-word perturbation (i.e. if a word contains multiple tokens, perturb them all at once) - this should be straightforward and give us a ~25% speed up.

In future we can look at n-grams etc, but I think you're right in that it's going to be pretty sensitive to grouping and so tricky to make reliable. Let's stick with words for now.

File: attribution/experiment_logger.py (hunk 5 above)
Comment (jessicarumbelow) [reply attribution/experiment_logger.py:149]: First option sounds good, thanks!

File: attribution/token_perturbation.py (hunk 4 above)
Comment (sebastian-sosa) [reply attribution/token_perturbation.py:79]: Sounds good! Yeah, it's easy to get the nearest, furthest and k-th furthest token with the provided functions. 
One thing to note though, is that the most similar tokens are usually variations of the same token, with different capitalization:

image

File: research/api_llm_attribution.ipynb
Comment (sebastian-sosa) [reply research/api_llm_attribution.ipynb:1]: time logging results here: https://leap-labs-workspace.slack.com/archives/C05P4FU88LW/p1714571954425999

File: attribution/token_perturbation.py
Comment (sebastian-sosa) [reply attribution/token_perturbation.py:1]: Thoughts on how to implement this: https://leap-labs-workspace.slack.com/archives/C05P4FU88LW/p1714662611730529

File: attribution/attribution_metrics.py (hunk 2 above)
Comment (sebastian-sosa) [reply attribution/attribution_metrics.py:10]: Okay, sounds good!

**Comment by jessicarumbelow** [comment]
image

Display bug again I think, seems to affect cosine only. I've replicated and pushed it - @sebastian-sosa can you take a look?

**Comment by jessicarumbelow** [comment]
Note that the column labels should be '13 minutes.', not 'the clock shows'

**Comment by jessicarumbelow** [comment]
Also, could we make perturb words or perturb tokens an optional argument, so it's easier for us to experiment?

**Comment by sebastian-sosa** [comment]
> Also, could we make perturb words or perturb tokens an optional argument, so it's easier for us to experiment?

This wasn't trivial but is now implemented. `perturb_word_wise` is a bool flag that allows you to control that, these were the changes made:
1. We split the input into "units", which can be either words or single tokens, and perturb all elements of a single unit on each step.
2. We made sure that the tokenization splits the input sentence into the same tokens no matter if we tokenize the text directly, or if we first split the text into words and then tokenize each word separately. To make this happen, the tokenizer is initialized with `add_prefix_space=True`  and `tokenizer.encode` uses the flag `add_special_tokens=False`
3. We store the flag `perturb_word_wise` for each experiment in the logger.

**Comment by sebastian-sosa** [comment]
> Display bug again I think, seems to affect cosine only. I've replicated and pushed it - @sebastian-sosa can you take a look?
> 
> Note that the column labels should be '13 minutes.', not 'the clock shows'

This is now fixed!

**Review by jessicarumbelow — APPROVED** [review]
